pkg/client/proxy.go

```diff
@@ -61,15 +65,19 @@ type ProxyClient struct {
 	Version int
 }
 
-func NewProxyClient(ctx context.Context, ctxCancel context.CancelFunc, address string, port int) (*ProxyClient, error) {
+func NewProxyClient(ctx context.Context, ctxCancel context.CancelFunc, address string, port int, tlsConfig *tls.Config) (*ProxyClient, error) {
 	getServiceCtx := func(serviceUrl string) (ServiceContext, error) {
 		dialOptions := []grpc.DialOption{
-			grpc.WithTransportCredentials(insecure.NewCredentials()),
 			grpc.WithKeepaliveParams(keepalive.ClientParameters{
 				Time:                time.Second * 10,
 				PermitWithoutStream: true,
 			}),
 		}
+		if tlsConfig != nil {
+			dialOptions = append(dialOptions, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+		} else {
+			dialOptions = append(dialOptions, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		}
```

Comment:
All other clients use the `util.Connect` helper method (now) to handle this. I prefer to do so here as well, unless there is a strong reason not to. It looks like the main differences are:

- We need to handle some strangeness around the missing protocol (`tcp://`) that `util.Connect` expects.
- `util.Connect` uses a 30 second keepalive instead of the 10 second keepalive established by @c3y1huang in https://github.com/longhorn/longhorn-instance-manager/pull/119. I suspect the 30 second timeout is fine here, but if not, we could make it configurable in `util.Connect`.

Reply:
Updated! It looks like `uti.Connect` should work with a raw `host:port` address since `parseEndoint` has a fallback when there's no `://` in the input.

Reply:
@c3y1huang 
Can you elaborate more on the increase of the 30-seconds keepalive ping time? The default value is 10 seconds.

Reply:
I don't recall why I used 10 seconds, might just be something that looks fitting at the time. Using 30 seconds keepalive looks fine to me.

Reply:
> I don't recall why I used 10 seconds, might just be something that looks fitting at the time. Using 30 seconds keepalive looks fine to me.

You use 30 seconds, but the default value is 10 seconds. Just curious about why you increased it to 30 seconds.

Reply:
To clarify (unless I'm missing something), @joshimoo made the decision to go with 30 seconds in https://github.com/longhorn/longhorn-instance-manager/pull/109 for other services. @c3y1huang chose 10 seconds (the minimum) in https://github.com/longhorn/longhorn-instance-manager/pull/119 for the proxy service.

Reply:
> To clarify (unless I'm missing something), @joshimoo made the decision to go with 30 seconds in #109 for other services. @c3y1huang chose 10 seconds (the minimum) in #119 for the proxy service.

Ah, yes. Sorry for the misunderstanding. BTW, any pros and cons for the bigger/smaller value?

Reply:
I am not too sure. If I understand correctly, a smaller value can help us detect a broken connection sooner during a long RPC (especially a streaming one). Since the server might not send back any data for a long time, the keepalive helps us know whether it is still trying.

On the other hand, the gRPC docs mention to avoid configuring keepalive much below one minute for DDoS reasons. That seems unlikely to be a major consideration for us since there are a limited number of clients.

https://github.com/longhorn/longhorn-instance-manager/blob/a0c0fa299379f9a9783ff5512c90a7defdac516c/vendor/google.golang.org/grpc/clientconn.go#L194-L210